% !TEX encoding = UTF-8 Unicode
\documentclass[11pt, a4paper]{article}
\usepackage{authblk}
\linespread{1.2}
\usepackage{palatino}
\usepackage{parskip}
\usepackage{graphicx}
\usepackage{hyperref}


% Title
\title{Manual for tsRFinder}
\author{Qinhu Wang}
\author{Weixing Shan\thanks{Email: wxshan@nwafu.edu.cn}}
\affil{Northwest A\&F University}
\date{\today}


%%% Begin document
\begin{document}

\maketitle
\tableofcontents

\section{Introduction}

Small RNAs are key regulators of gene expression, such as miRNA, siRNA, and piRNA. The tRNA-derived small RNA (tsRNA; also tRNA-derived RNA fragments, tRF), is a class of novel small RNA have been identified recently. However, there is no public tool available for tsRNA prediction yet. We thus developed tsRFinder for tsRNA prediction. It takes the raw data of small RNA sequencing reads and the reference genome sequence, and identifies tsRNA with additional sequence and statistical analysis for you automatically.

\section{How to install}

\subsection{Dependencies}

tsRFinder depends on the following programmes, please check and install them \footnote{The dependencies version is based on the oldest test enviroment we have.} at first:

\begin{itemize}

\item Perl, greater than v5.10.1, required, for tsRFinder.pl execution. And it is always installed already in most of the UNIX-like operating systems.
Two perl modules, Config::Simple and Getopt::Std (always build-in) are need to install if you havn't.
\item bowtie, greater than v1.0.0, required, for small RNA mapping. \\https://github.com/BenLangmead/bowtie
\item R, greater than v2.15.2, required, for small RNA data analysis and illustration. \\http://www.r-project.org
\item tRNAscan-SE, greater than v1.3.1, optional if you want to prepare tRNA input yourself, for tRNA prediction. \\http://lowelab.ucsc.edu/tRNAscan-SE
\item fastx\_toolkit, greater than v0.0.14, optional if you have already processed the raw sequencing data yourself. \\https://github.com/agordon/fastx\_toolkit

\end{itemize}

\subsection{Installation}

tsRFinder is maintained on GitHub and is ready-to-use, no compilation is required. However, if you take some time to improve the configuration, it may save you a lot of time for trouble shooting.

First, you can clone \footnote{If git is not installed, download it from http://git-scm.com} tsRFinder by typing:

{\small \begin{verbatim}
git clone https://github.com/wangqinhu/tsRFinder.git
\end{verbatim}}

in the terminal. Alternatively, you can download it from

{\small \begin{verbatim}
https://github.com/wangqinhu/tsRFinder/releases/latest
\end{verbatim}
}

and then decompress the zip file or the tarball.

When tsRFinder is cloned or unpacked, move the entire directory to an proper place and add the tsRFinder path to the environment setting. For example, if tsRFinder is placed in /your/path/of/tsRFinder, then type the following in the terminal if you are using bash.

{\small \begin{verbatim}
echo export tsR_dir="/your/path/of/tsRFinder" >> ~/.bashrc
source ~/.bashrc
\end{verbatim}}

And now, you can run tsRFinder for your dataset.

\section{How to use}

\subsection{Preparation the dataset}

Before running tsRFinder, you are asked to prepare/download the following two files: (1) the reference genome sequence, or the reference tRNA sequence and, (2) the small RNA reads.

We strongly recommend you using the reference genome sequence and the raw small RNA sequencing data, since tsRFinder can help you prepare the reference tRNA data and clean small RNA data automatically. If you want to prepare the tRNA file and small RNA reads file by yourself, you can run the demo data (enable debug mode in command line option: \emph{-m debug}) and find out which exact format of tRNA reference and small RNA reads file can be accepted instead, this is allowed but not encouraged. You can also found the format description in the FAQ part of this document.

\subsection{Running the pipeline}

tsRFinder supplies two ways for arguments input, you can use both configuration file and command line option. We recommend you use a command line option for debugging and building your configuration file. Once your inputs have been determined, you can write it to a configuration file for your analysis.

If tsRFinder is properly installed, you can run tsRFinder from your terminal directly, see the usage below.

{\small \begin{verbatim}
tsRFinder usage:

    tsRFinder.pl <option>

    -c  Configuration file
    -l  Label
    -g  Reference genomic sequence
    -t  Reference tRNA sequence
    -s  Small RNA sequence
    -a  Adaptor sequence
    -n  Min read length            [defalut 18]
    -x  Max read length            [default 45]
    -f  Small RNA family threshold [default 72]
    -w  tRNA with/without label    [defualt no]
    -h  Help
    -v  Version

Example:

    tsRFinder.pl -c demo/tsR.conf
\end{verbatim}}

\section{Demo}

To have a quick but rough overview that how tsRFinder looks like, see the the animated gif demo in doc/demo.gif or online.

\subsection{Demo data}

\textbf{Demo refseq}: we used serval random sequences embedded with some real tRNAs as pseudo reference genome sequence. This small sequence data in FASTA format can be accessed in the file "tsRFinder/demo/genome.fa". In your analysis, if you have a reference genome, just replace it with the reference sequence; if you don't have reference genome sequence, you can use the reference tRNA sequence instead. Figure \ref{refseq} shows what a refseq file looks like.

\begin{figure}[htbp]
\begin{center}
\includegraphics[width=12cm]{refseq.png}
\caption{Screenshot of the reference sequence in fasta format} 
\label{refseq}
\end{center}
\end{figure}

\textbf{Demo sRNA}: we extracted a bit of raw reads from some real experimental data as a demo here. Each read of the raw small RNA data have 4 lines, just like what we have show in Figure \ref{fastq}.

\begin{figure}[htbp]
\begin{center}
\includegraphics[width=12cm]{fastq.png}
\caption{Screenshot of the small RNA sequence in fastq format} 
\label{fastq}
\end{center}
\end{figure}

\subsection{Demo running}

tsRFinder allowes you specify your inputs via a separate configuration file, for example, here is the content of our demo tsR.conf:

\begin{verbatim}
#===================================#
#                                   #
#    Configuration for tsRFinder    #
#                                   #
#===================================#

label               :  Abc
reference_genome    :  demo/genome.fa 
reference_tRNA      :
sRNA                :  demo/sRNA.fq
adaptor             :  TGGAATTCTCGGGTGCCAAGG
min_read_length     :  18
max_read_length     :  45
family_threshold    :  72
tRNA_with_label     :  no
\end{verbatim}

Currently we have 9 arguments used for filling. The argument items and the inputs are separated by colon (":"). We recommend you using the first three letters of the organism you are analysing as a label (e.g. for \textit{Arabidopsis thaliana} we use \textit{Ath}, \emph{-l} in command line option); the paths of reference genome (\emph{-g}) and small RNA (\emph{-s}) should be supplied at least. If you are using a raw sequence data, please also input the adaptor sequence (\emph{-a}). If you are not using a reference tRNA (\emph{-t}) prepared by yourself, leave this argument to EMPTY please.

Once your configuration file is well prepared, typing the following in the terminal to run tsRFinder. In this demo, the configuration file (\emph{-c}) is at demo/tsR.conf, so we write like this:

\begin{verbatim}
# We have set "tsR_dir" as an environment variable before
cd tsR_dir
./tsRFinder.pl -c demo/tsR.conf
\end{verbatim}

\subsection{Demo output}

By default, tsRFinder will give you a summary of which files have been outputted and some basic statistic. The predicted or user inputed tRNA sequence, the small RNA clean data, the tRNA reads, and the predicted tsRNA sequence were listed. Meanwhile, tsRFinder gives you additional summary on tRNA/tsRNA expression (including 5' tsRNA and 3' tsRNA), text map (tmap) of small RNA mapped to tRNA, tsRNA family, graphics showing the expression evaluated by small RNA data, and also the cleavage site. Additional, tsRFinder supplies a figure showing the small RNA and tRNA reads length distribution (Figure \ref{distribution}) for you.

See our demo summary here:

{\small \begin{verbatim}
---------
 SUMMARY 
---------

    tRNA seq : /Users/wangqinhu/tsRFinder/Abc/tRNA.fa
       Total : 5


  sRNA reads : /Users/wangqinhu/tsRFinder/Abc/sRNA.fa
       Total : 12432
      Unique : 7412

  tRNA reads : /Users/wangqinhu/tsRFinder/Abc/tRNA.read.fa
       Total : 286
      Unique : 52

   tsRNA seq : /Users/wangqinhu/tsRFinder/Abc/tsRNA.seq
       Total : 7
      Unique : 6

tsRNA report : /Users/wangqinhu/tsRFinder/Abc/tsRNA.report.xls

    text map : /Users/wangqinhu/tsRFinder/Abc/tsRNA.tmap

  visual map : /Users/wangqinhu/tsRFinder/Abc/images

distribution : /Users/wangqinhu/tsRFinder/Abc/distribution.pdf

    cleavage : /Users/wangqinhu/tsRFinder/Abc/cleavage.txt

tsRNA family : /Users/wangqinhu/tsRFinder/Abc/tsRNA.fam

stat. by BDI :
 Sensitivity : 0.94
 Specificity : 0.7838
    Accuracy : 0.8764

---------
\end{verbatim}
}

\begin{figure}[htbp]
\begin{center}
\includegraphics[width=12cm]{distribution.pdf}
\caption{Small RNA and tRNA reads distribution}
\label{distribution}
\end{center}
\end{figure}

\subsection{Visualization of tmap data}

To examine the map of tsRNA, we developed a vim syntax plugin for visualization (Figure \ref{tmap}). If you want to enable color text map, copy lib/tmap.vim into your vim syntax folder, and put the following line into your .vimrc file:

\begin{verbatim}
au BufNewFile,BufRead *.tmap  setf tmap
\end{verbatim}

When tmap.vim is correctly installed, open the tsRNA.tmap file with vim you will obtain a color text map, like this:

\begin{verbatim}
vim tsRNA.tmap
\end{verbatim}

\begin{figure}[htbp]
\begin{center}
\includegraphics[width=12cm]{tmap.png}
\caption{Screenshot of color tmap} 
\label{tmap}
\end{center}
\end{figure}

If you prefer plain text view without highlighting, just open tsRNA.tmap with any kind of text editors you have.

\section{FAQ}

\textbf{1. Can tsRFinder run on Windows?}

No. tsRFinder used some build-in program of UNIX-like systems, for example, awk, grep and head, thus running on Windows may lead unexpected errors, we strongly recommend you running tsRFinder on Linux or OS X.

\textbf{2. What's the length required for small RNA reads?}

We recommend you sequencing from 15 - 50 nt for small RNA, however, 18 - 30 nt is OK if your tsRNA is shorter than 30 nt (such as tRF).

\textbf{3. What's the tRNA format required for input?}

At first we encourage you using a reference genomic sequence to instead it, since tsRFinder will help you extract the tRNA sequence via tRNAscan-SE. And if you really want to use a file prepared yourself, please follow this format (see example): first line, begin with "$>$", followed by label "$Abc$", then which tRNA "$tRNA-ProCGG1$";  second line, the sequence of the tRNA; and third line, the secondary structure of tRNA, "$><$" and "$()$" both were acceptable.

{\scriptsize \begin{verbatim}
tRNA example:
>AbctRNA-ProCGG1
GGCCTCGTGGTCTAGTGGTATGATTCTCGCTTCGGGTGCGAGAGGtCCCGGGTTCGATTCCCGGTGAGGCCC
>>>>>>>..>>>.........<<<.>>>>>.......<<<<<.....>>>>>.......<<<<<<<<<<<<.
\end{verbatim}}

\textbf{4. What's the sRNA format required for input?}

We recommend you using a raw data. To have an sRNA file for input prepared yourself, please follow this format (see example): first line, begin with "$>$", then the label "$Abc$", then the 7\-bit index, followed by "$\_$" (or "$|$", "$-$" and white space) and the read number; second line, the sequence of the read.

{\scriptsize \begin{verbatim}
sRNA example:
>Abc0000001_772
CAGGTGGTCAGGTAGAGAATACCAAGGCGCT
>Abc0000002_475
AGGTGGTCAGGTAGAGAATACCAAGGCGCT
\end{verbatim}}


\textbf{5. I have problem in installing tsRFinder and/or the dependencies, where to get more help?}

You can go to the offical support website or create new issue for tsRFinder repository on GitHub. The URL is https://github.com/wangqinhu/tsRFinder/issues/new

\textbf{6. Where to report bugs?}

Goto https://github.com/wangqinhu/tsRFinder/issues/new

\textbf{7. Can we use tsRFinder for commercial purpose?}

Yes. tsRFinder is free, open source software, see the MIT license.

\textbf{8. The 'nwalign' is not suit for my operating system, what can I do?}

tsRFinder using Needleman-Wunsch algorithorim nwalign for small RNA alignment, we pre-build the binaries for some of the recently OS X / Linux, if you find it not suitable for your system or want to complile it by your self, goto lib/src directory and type 'make' to build \footnote{The gcc compiler is required.} it.

\end{document}
